### DISCUSSION TITLE
TryThenThrow debates

### DISCUSSION THREAD

**Discussion by HamletHakobyan** [vkhorikov/CSharpFunctionalExtensions#499] [Ideas]
### Preamble
Maybe something like this is exists in this repo but, unfortunately, I didn't find.

### Motivation

In some systems with global exception handling when unexpected exceptions throw it make sense to leave it for handling on top level of stack. In case when we have possibility of both expected and unexpected exception maybe make sense to have such an overload of `Try`

```
public static async Task<Result<T>> Try<T>(Func<Task<T>> func, Func<Exception, (string, bool)> errorHandler)
{
    try
    {
        var result = await func();
        return Success(result);
    }
    catch (Exception exc)
    {
        var (message, isHandled) = errorHandler(exc);
        if (!isHandled) throw;

        return Failure<T>(message);
    }
}
```
and a use case

```
var result = await Result.Try(
    async () => await _someService.SomeMethodAsync(),
    ex => ex is HttpRequestException { StatusCode: HttpStatusCode.NotFound } reqEx ? (reqEx.Message , true) : ("", false)
);
```

**Comment by vkhorikov** [comment]
I don't think we have such an extension. Looks good to me, feel free to submit a PR.
